pkg/networkserver/grpc_asns.go

```diff
@@ -245,8 +246,13 @@ func matchQueuedApplicationDownlinks(ctx context.Context, dev *ttnpb.EndDevice,
 	return nil
 }
 
+var errDownlinkQueueCapacityExceeded = errors.DefineResourceExhausted("downlink_queue_capacity_exceeded", "Downlink queue capacity exceeded")
+
 // DownlinkQueueReplace is called by the Application Server to completely replace the downlink queue for a device.
 func (ns *NetworkServer) DownlinkQueueReplace(ctx context.Context, req *ttnpb.DownlinkQueueRequest) (*pbtypes.Empty, error) {
+	if n := len(req.Downlinks); uint64(n) > ns.downlinkQueueCapacity {
```

Comment:
In theory this is `2 * ns.downlinkQueueCapacity`, since we may push both the `session` and the `pending_session` at the same time, correct ?

Reply:
Yep, I was making this change while you were reviewing https://github.com/TheThingsNetwork/lorawan-stack/pull/3236/commits/ad7ef7ccaafd952bb305c6b0e56b796334ec51a3